Simplify Logger implementation

AndrejMitrovic opened the pull request:
At first I had a strange error where one module was complaining that `Logger` was implemented in multiple modules. There was an `AddLogger` mixin in both modules. Then I noticed the `mixed-in` Logger was public. Well, it could just be one symbol instead of having a separate instance per module.

AndrejMitrovic committed:
Simplify Logger implementation

Previously the AddLogger mixin was declaring a separate
Logger type for each module where AddLogger!() was
mixed-in.

Additionally removed the need to import the ocean
logger.

Also made TransactionPool use our nothrow Logger
rather than the Ocean one directly.

Geod24 commented on source/agora/common/TransactionPool.d:
```diff
@@ -39,6 +38,8 @@ version (unittest)
     import agora.utils.Test;
 }
 
+mixin AddLogger!();
```
Can you add a comment, e.g.:
```suggestion
// Note: This needs to be before the call to `TransactionPool.initialize`
mixin AddLogger!();
```

Geod24 commented on source/agora/utils/Log.d:
```diff
@@ -31,86 +31,98 @@ import ocean.transition;
+    // workaround: weird CT bug in NetworkManager.d call:
+    // log.info("Discovery reached. {} peers connected.", this.peers.length);
+    // source/agora/network/NetworkManager.d(149,12): Error: no property info for type Logger
+    public void info (Args...) (Args args) @safe nothrow
```
This bug has been fixed no ?

AndrejMitrovic replied:
It's funny you mention this, because I seem to have other issues now too. Like calling `.fatal` fails.

I feel like `opDispatch` is a very hacky implementation in DMD..

Geod24 reviewed (approved):
Nice

AndrejMitrovic committed:
Logger: Remove .info() workaround

AndrejMitrovic committed:
Logger: Remove attributes from opDispatch()

These can be inferred because it's a template.

AndrejMitrovic commented:
Thanks. I've removed the `info()` call. And I've also removed attributes from `opDispatch` as those can be inferred. I think forcing those attributes makes DMD gag the error messages when instantiating it, so we get errors like "log.fatal not found".

Geod24 commented:
CircleCI is complaining

AndrejMitrovic commented:
> CircleCI is complaining

Looking into it.

AndrejMitrovic commented:
Very curious.

```
Thread 2 received signal SIGSEGV, Segmentation fault.
0x0000000100002af0 in _D5ocean4util3log6LoggerQh__T6formatTAyaZQmMFNfEQBuQBrQBp7ILoggerQi5LevelAxaQBnZv (this=0x0, level=ocean.util.log.ILogger.ILogger.Info, fmt=...,
    _param_2=...) at /Users/andrejmitrovic/dev/agora/submodules/ocean/src/ocean/util/log/Logger.d:866
866	            if (this.buffer_.length)
(gdb) bt
#0  0x0000000100002af0 in _D5ocean4util3log6LoggerQh__T6formatTAyaZQmMFNfEQBuQBrQBp7ILoggerQi5LevelAxaQBnZv (this=0x0, level=ocean.util.log.ILogger.ILogger.Info,
    fmt=..., _param_2=...) at /Users/andrejmitrovic/dev/agora/submodules/ocean/src/ocean/util/log/Logger.d:866
#1  0x0000000100002a7b in _D5ocean4util3log6LoggerQh__T4infoTAyaZQkMFNfAxaQnZv (this=0x0, fmt=..., _param_1=...)
    at /Users/andrejmitrovic/dev/agora/submodules/ocean/src/ocean/util/log/Logger.d:505
#2  0x0000000100057fa0 in _D5agora5utils3Log6Logger__T10opDispatchVAyaa4_696e666fTQpTQsZQBiMFNbNfQBeQBhZv (this=<optimized out>, _param_0=..., _param_1=...)
    at /Users/andrejmitrovic/dev/agora/source/agora/utils/Log.d-mixin-66:66
#3  0x0000000100065549 in _D5agora6common15TransactionPoolQr6__ctorMFxAyaZCQBvQBsQBoQBr (this=0x102f1ab20, db_path=...)
    at /Users/andrejmitrovic/dev/agora/source/agora/common/TransactionPool.d:103
#4  0x000000010005d596 in _D5agora4node8FullNodeQj7getPoolMFAyaZCQBl6common15TransactionPoolQr (this=0x102f20990, data_dir=...)
    at /Users/andrejmitrovic/dev/agora/source/agora/node/FullNode.d:325
#5  0x000000010005b407 in _D5agora4node8FullNodeQj6__ctorMFxSQBh6common6ConfigQhDFNbNeZvZCQCkQChQCfQCi (this=0x102f20990, config=..., onValidatorsChanged=...)
    at /Users/andrejmitrovic/dev/agora/source/agora/node/FullNode.d:139
#6  0x000000010005af7c in _D5agora4node9ValidatorQk6__ctorMFxSQBi6common6ConfigQhZCQCdQCaQByQCb (this=0x102f20990, config=...)
    at /Users/andrejmitrovic/dev/agora/source/agora/node/Validator.d:69
#7  0x000000010005abb0 in _D5agora4node6Runner7runNodeFSQBc6common6ConfigQhZCQBxQBu8FullNodeQj (config=...)
    at /Users/andrejmitrovic/dev/agora/source/agora/node/Runner.d:59
#8  0x000000010005a939 in _D5agora4node4mainQfFAAyaZ9__lambda2MFZCQBmQBj8FullNodeQj () at /Users/andrejmitrovic/dev/agora/source/agora/node/main.d:75
#9  0x00000001001bf04c in _D4vibe4core4task12TaskFuncInfo__T3setTDFZC5agora4node8FullNodeQjZQBgMFKQBhZ12callDelegateFKSQDnQDlQDjQDhZv (tfi=...)
    at /Users/andrejmitrovic/dev/agora/submodules/vibe-core/source/vibe/core/task.d:702
#10 0x0000000100595ca2 in vibe.core.task.TaskFiber.shutdown().__lambda2() () at task.d:518
#11 0x00000001005954e4 in vibe.core.task.TaskFiber.run() (this=0x102f1ea00) at task.d:482
#12 0x0000000100769039 in ?? ()
#13 0x0000000000000000 in ?? ()
```

AndrejMitrovic commented:
@Geod24 I found a new issue. I'm not sure whether this is by design, by if it is then it's by dumb design. If you have this:

```
mixin ModuleCtor!();

static this ()
{
    ....
}
```

Then the second module constructor will run first, followed by the mixed-in one.

Geod24 commented:
Commits and messages are a bit effed up here.

AndrejMitrovic commented:
Ah right it looks like I edited the wrong commit.

Geod24 commented:
> Previously the AddLogger mixin was declaring a separate `Logger` type for each module where `AddLogger!()` was mixed-in.

Ouch, I understand now...
